### PR TITLE
Fixing shorewall

### PR DESCRIPTION
**Pull request by kaspervd**
See discussion at https://github.com/HiSPARC/publicdb/issues/159

### CODE REVIEW

**Comment by davidfokkema** [comment]
I can't build the VM at the moment because of DNS errors. It seems that shorewall(?) overwrites the /etc/resolv.conf file. Can you build the VM from scratch?

**Comment by kaspervd** [comment]
Yes, this works in my case

**Comment by davidfokkema** [comment]
@kaspervd Can you post the resulting /etc/resolv.conf?

**Comment by kaspervd** [comment]
That would be

```
# Generated by NetworkManager
search localdomain
nameserver 10.0.2.3
```

File: provisioning/shorewall-firewall.yml [hunk 1]
```diff
@@ -1,13 +1,26 @@
 ---
-- name: install shorewall firewall
-  yum: name=shorewall
+- name: ensure local source directory exists
+  file: path=/usr/local/src/hisparc state=directory
+        owner=hisparc group=hisparc mode=2775
+  become: true
+
```
Comment (davidfokkema) [inline provisioning/shorewall-firewall.yml]: We no longer need to ensure the source directory exists, right?

**Comment by tomkooij** [comment]
I can't build the vpn VM on Windows: I use the `ansible_local` provisioner. This runs ansible on the vpn VM. This fails when shorewall is activated: Packages can no longer be downloaded. 

As long as we use ansible from another host this is no issue, but I cannot easily test on my Windows laptop.

**Comment by davidfokkema** [comment]
@tomkooij I'm not sure if this is a problem because of ``ansible_local``. I have the same symptoms because the DNS servers are unreachable from outside Nikhef. Did you test at Nikhef, or somewhere else? If somewhere else, it would be helpful if you could try again on Monday.

**Comment by tomkooij** [comment]
Will test on Monday!

**Comment by tomkooij** [comment]
I had some time to fiddle with this today. Provisioning stops when `/etc/resolv.conf` is copied in `openvpn.yml`. Putting `openvpn.yml` as the last step in `common.yml` solves it. But that is not a lasting solution.

f8bf319 is a bit of a hack. This does not copy the file but only triggers the handler to copy the file at the end of provisioning. This *always* triggers that handler. IMO we cannot rely on letting some other task trigger this handler, because `/etc/resolv.conf` may be altered regardless of such task running.

**Comment by tomkooij** [comment]
85bc78b is (IMO) a better way to handle this: It adds `post_tasks` that run at the end of provisioning. This task copies the `resolv.conf` files. This makes sure the DNS is altered after everything is done. (I don't think service have to be restated after changing `resolv.conf`. I checked `dnsmasq` and that polls `resolv.conf` for changes).

This still builds the vpn VM on my Windows laptop.

**Comment by davidfokkema** [comment]
I'm not sure why, but when I first build the VM, the resolv.conf ends up containing NetworkManager-related autodetected stuff. When I again provision the VM, it correctly ends up with:
```
search nikhef.nl his
nameserver 127.0.0.1
```
When I reboot the VM, it again ends up with NetworkManager-related stuff. It gets its DNS server from the DHCP request, I think. It is important that the nameserver points to itself (dnsmasq). So unfortunately, I think this has to be looked into.

**Comment by davidfokkema** [comment]
I dismissed my own review, but now I'm not listed as a reviewer anymore. That was not my intention, sorry.

**Comment by tomkooij** [comment]
Mmmh... This looks like the `NetworkManger` default behaviour ([man-page](http://manpages.ubuntu.com/manpages/zesty/man5/NetworkManager.conf.5.html)) 
```
 dns
           Set the DNS (resolv.conf) processing mode.

           default: NetworkManager will update resolv.conf to reflect the
           nameservers provided by currently active connections. This is the
           default if the key is not specified, unless the system is
           configured to use systemd-resolved; in this case the default is
           systemd-resolved

           dnsmasq: NetworkManager will run dnsmasq as a local caching
           nameserver, using a "split DNS" configuration if you are connected
           to a VPN, and then update resolv.conf to point to the local
           nameserver.

           unbound: NetworkManager will talk to unbound and dnssec-triggerd,
           providing a "split DNS" configuration with DNSSEC support. The
           /etc/resolv.conf will be managed by dnssec-trigger daemon.

           systemd-resolved: NetworkManager will push the DNS configuration to
           systemd-resolved

           none: NetworkManager will not modify resolv.conf. This implies
           rc-manager unmanaged

```

Since we are running our own `dnsmasq` think we need to add `dns=none` to `NetworkManager.conf`.

If we need NetworkManager to run dnsmasq, things are more complicated :-)

**Comment by davidfokkema** [comment]
We've never seen this before. Did the default change, or did CentOS not use NetworkManager previously?

**Comment by tomkooij** [comment]
@davidfokkema NetworkManager is running on tietar. But there is no `etc/NetworkManager/Networkmanager.conf`

tietar does use networkmanager:
```
170306 10:30:53 hisparc@tietar /etc/sysconfig/network-scripts
> cat ifcfg-em1
[...]
NM_CONTROLLED="yes"
```

The default configuration of NetworkManager may have changed but I am not aware of the "last known good" configuration.

**Comment by davidfokkema** [comment]
I checked on Tietar itself. All interfaces are NetworkManager-controlled, and I see no config file...

**Comment by tomkooij** [comment]
`/etc/resolv.conf` gets overwritten on my VM even if no configfile exists. I guess we need to add `dns=none` to the config to fix this.

**Comment by tomkooij** [comment]
This is *very* annoying.

Adding `dns=none` to `/etc/NetworkManager/NetworkManger.conf` does not work.
Adding `PEERDNS=no` to `/etc/sysconfig/network-scripts/if-eth0` does not work. (Not really a solution, because those scripts are network interface dependent)
Both command should tell `NetworkManager` to not update `/etc/resolv.conf` sadly, this is ignored and `/etc/resolv.conf` get overwritten at reboot (or more precisely at DHCP after `ifup eth0`)

If I completely disable NetworkManager:
```
[hisparc@vpn ~]$ sudo chkconfig NetworkManager off
```
`/etc/resolv.conf` gets overwritten by `dhclient`:
```
[hisparc@vpn ~]$ cat /etc/resolv.conf
; generated by /sbin/dhclient-script
nameserver 10.0.2.3
```
This can be fixed as well... [(snippet)](http://unix.stackexchange.com/questions/174349/what-overwrites-etc-resolv-conf-on-every-boot/188464#188464) But... *sigh*.

Making `/etc/resolv.conf` **immuttable** does seem to work:
```
[hisparc@vpn ~]$ sudo chattr +i /etc/resolv.conf
```
But that is a major hack. This does seems to protect against NetworkManager and `dhclient-script`.

I'm not sure what to do, so I'll leave this as-is for now.

**Comment by davidfokkema** [comment]
Not sure if this is relevant, but the configuration file is misspelled in your comment. Manger instead of manager. Major headache this is...

**Comment by tomkooij** [comment]
Just a typo. I'm quite sure I used the correct configuration file. 

Anyway, some solutions:
 - Overwriting `/etc/resolv.conf` while provisioning and useing `chattr +i /etc/resolv.conf` to stop all others touching it. 
 - provision `/etc/dhcp/dhclient-XXX.conf` instead of provisioning `/etc/resolv.conf'
 
The first is an easy hack, the second will require some careful ansible scripts to provision the actual interface. 

Headache. Yes.

**Comment by davidfokkema** [comment]
Pity. Is the interface really named eth0? Or is it en0?

**Comment by tomkooij** [comment]
In my VM the interface is named eth0.
```
[root@vpn network-scripts]# ifconfig
eth0      Link encap:Ethernet  HWaddr 08:00:27:4B:A4:28
          inet addr:10.0.2.15  Bcast:10.0.2.255  Mask:255.255.255.0
(...)
```

This is because Vagrant forces this name (I think):
`/etc/sysconfig/network-scripts/ifcfg-eth0`:
```
[root@vpn network-scripts]# cat ifcfg-eth0
DEVICE="eth0"
(...)
```

**Comment by tomkooij** [comment]
5340a28 solves NetworkManager messing up /etc/resolv.conf when configuring interfaces using DHCP.
This should only affect VMs, since the real tietar does not use DHCP.

Outside of nikhef provisioning will only work the first time. Once /etc/resolv.conf is changed, DNS is unreachable.

**Comment by 153957** [comment]
Shorewall is in EPEL for CentOS 7:
http://dl.fedoraproject.org/pub/epel/7/x86_64/Packages/s/

**Comment by tomkooij** [comment]
Superseded by #230

**Pull request closed without merging**